frontend: Refactor main window menus

cg2121 opened the pull request:
### Description
This makes it so we don't have to manually delete each menu each time they are shown, they are now created on demand, saving some memory. This also moves the menu code to a separate file.

### Motivation and Context
The menu code has always bothered me

### How Has This Been Tested?
Clicked through each menu to made sure they still worked

### Types of changes
- Code cleanup (non-breaking change which makes code smaller or more readable)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

tytan652 commented on UI/window-basic-main-menus.cpp:
```diff
@@ -0,0 +1,234 @@
+
+void OBSBasic::SetDeinterlacingMode()
+{
+	QAction *action = reinterpret_cast<QAction *>(sender());
```
`sender()` is a last resort solution provided by Qt which does not respect POO programming.

I don't think it's a good idea to heavily rely on it.

cg2121 replied:
This PR just copies the code from OBSBasic, so nothing has changed. We can always change the sender() code in a future PR.

cg2121 committed:
frontend: Refactor main window menus

This makes it so we don't have to manually delete each menu each time
they are shown, they are now created on demand, saving some memory.

cg2121 commented:
Updated to adapt to the new frontend structure.

Warchamp7 commented on frontend/forms/OBSBasic.ui:
```diff
@@ -928,7 +923,6 @@
     <addaction name="separator"/>
     <addaction name="stats"/>
     <addaction name="separator"/>
-    <addaction name="multiviewProjectorMenu"/>
```
Qt will be 'smart' about not showing multiple separators.

Warchamp7 commented on frontend/widgets/OBSBasic.hpp:
```diff
@@ -931,31 +923,27 @@ public slots:
 			action->setProperty("monitor", i);
 		}
+
+		return menu;
```
It should require a parent passed into the function, or always parent them to OBSBasic itself.

cg2121 replied:
Updated to include parent parameter.

Warchamp7 commented on frontend/widgets/OBSBasic_MainControls.cpp:
```diff
@@ -681,3 +682,11 @@ void OBSBasic::on_OBSBasic_customContextMenuRequested(const QPoint &pos)
+{
+	multiviewMenu.reset(CreateProjectorMenu(QTStr("Projector.Open.Multiview"), nullptr, ProjectorType::Multiview));
+
+	QAction *separatorAction = ui->viewMenu->actions().at(ui->viewMenu->actions().count() - 2);
```
This is very cursed and we should find a better way to do it.

Should in theory be able to insert the new menu before the old one, then remove the old one.

cg2121 replied:
This wouldn't work when OBS is first loaded because the menu is not created yet.

Warchamp7 replied:
It would work if there were a placeholder entry within the UI form still, no? Otherwise what you're saying wouldn't work now either.

Warchamp7 commented on frontend/widgets/OBSBasic.hpp:
```diff
@@ -931,31 +923,27 @@ public slots:
 	static QList<QString> GetProjectorMenuMonitorsFormatted();
 	template<typename Receiver, typename... Args>
-	static void AddProjectorMenuMonitors(QMenu *parent, Receiver *target, void (Receiver::*slot)(Args...))
+	static QMenu *CreateProjectorMenu(const QString &name, Receiver *target, void (Receiver::*slot)(Args...))
```
Update to camelCase

cg2121 replied:
Fixed

Warchamp7 commented on frontend/widgets/OBSBasic_SceneItems.cpp:
```diff
@@ -362,157 +363,245 @@ void OBSBasic::ResetAudioDevice(const char *sourceId, const char *deviceId, cons
-	ADD_ORDER("TopFieldFirst", OBS_DEINTERLACE_FIELD_ORDER_TOP);
-	ADD_ORDER("BottomFieldFirst", OBS_DEINTERLACE_FIELD_ORDER_BOTTOM);
-#undef ADD_ORDER
+	addOrder(QTStr("TopFieldFirst"), OBS_DEINTERLACE_FIELD_ORDER_TOP);
```
```suggestion
	addOrder(QTStr("Deinterlacing.TopFieldFirst"), OBS_DEINTERLACE_FIELD_ORDER_TOP);
```

cg2121 replied:
Fixed

Warchamp7 commented on frontend/widgets/OBSBasic_SceneItems.cpp:
```diff
@@ -362,157 +363,245 @@ void OBSBasic::ResetAudioDevice(const char *sourceId, const char *deviceId, cons
-	ADD_ORDER("BottomFieldFirst", OBS_DEINTERLACE_FIELD_ORDER_BOTTOM);
-#undef ADD_ORDER
+	addOrder(QTStr("TopFieldFirst"), OBS_DEINTERLACE_FIELD_ORDER_TOP);
+	addOrder(QTStr("BottomFieldFirst"), OBS_DEINTERLACE_FIELD_ORDER_BOTTOM);
```
```suggestion
	addOrder(QTStr("Deinterlacing.BottomFieldFirst"), OBS_DEINTERLACE_FIELD_ORDER_BOTTOM);
```

cg2121 replied:
Fixed

Warchamp7 reviewed (changes requested):
Did some quick testing and menus all seemed to behave correctly still.

Left a few comments, and just making a general note here to have any of the new/renamed functions be camelCase instead of PascalCase.
